pageserver/src/page_service.rs

```diff
@@ -849,67 +849,47 @@ impl PageServerHandler {
     /// the LSN that should be used to look up the page versions.
     async fn wait_or_get_last_lsn(
         timeline: &Timeline,
-        mut lsn: Lsn,
-        latest: bool,
+        lsn: Lsn,
+        horizon: Lsn,
         latest_gc_cutoff_lsn: &RcuReadGuard<Lsn>,
         ctx: &RequestContext,
     ) -> Result<Lsn, PageStreamError> {
-        if latest {
-            // Latest page version was requested. If LSN is given, it is a hint
-            // to the page server that there have been no modifications to the
-            // page after that LSN. If we haven't received WAL up to that point,
-            // wait until it arrives.
-            let last_record_lsn = timeline.get_last_record_lsn();
-
-            // Note: this covers the special case that lsn == Lsn(0). That
-            // special case means "return the latest version whatever it is",
-            // and it's used for bootstrapping purposes, when the page server is
-            // connected directly to the compute node. That is needed because
-            // when you connect to the compute node, to receive the WAL, the
-            // walsender process will do a look up in the pg_authid catalog
-            // table for authentication. That poses a deadlock problem: the
-            // catalog table lookup will send a GetPage request, but the GetPage
-            // request will block in the page server because the recent WAL
-            // hasn't been received yet, and it cannot be received until the
-            // walsender completes the authentication and starts streaming the
-            // WAL.
-            if lsn <= last_record_lsn {
-                lsn = last_record_lsn;
-            } else {
-                timeline
-                    .wait_lsn(
-                        lsn,
-                        crate::tenant::timeline::WaitLsnWaiter::PageService,
-                        ctx,
-                    )
-                    .await?;
-                // Since we waited for 'lsn' to arrive, that is now the last
-                // record LSN. (Or close enough for our purposes; the
-                // last-record LSN can advance immediately after we return
-                // anyway)
-            }
+        let last_record_lsn = timeline.get_last_record_lsn();
+        // Horizon = 0 (INVALID) is treated as LSN interval degenerated to point [lsn,lsn].
+        // It as done mostly for convenience (because such get_page commands are widely used in tests) and
```

Comment:
> Why it is hard for them to use horizon=lsn?

I don't know best terms, but compute need to specify:
1. "desired" LSN - which LSN compute expects to be acceptable
2. upper boundary LSN - PS should not send page with greater LSN to preserver consistency

If `latest=true` (primary node), then upper boundary is always +inf: nobody except this node can produce more recent version of the page.
In case of replica it is not true: replica can lag from primary node and PS and should not receive pages newer than its `last_replay_lsn`. But it is not good always to request pages at  `last_replay_lsn` because replica can be **ahead** PS and so it has to wait until PS caught up (while for this partitucular page it is not needed).

So actually we need to handle just three cases:
1. [page_last_written_lsn, +inf] - primary node
2. [page_last_written_lsn, last_replay_lsn] - hot-standby replica (receiving WAL from primary)
3. [snapshot_lsn, shapshot_lsn] - static RO replica (not receiving WAL fro primary)

Case [0, lsn] is not actually needed and added mostly for convenience as alias for [lsn,lsn]